core/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java

```diff
@@ -95,7 +104,22 @@ public InternalAggregation buildAggregation(long zeroBucket) throws IOException
             final CompositeValuesSource.Collector collector =
                 array.getLeafCollector(context.ctx, getSecondPassCollector(context.subCollector));
             int docID;
+            DocIdSetIterator docIt = null;
+            if (needsScores()) {
+                Scorer scorer = weight.scorer(context.ctx);
+                // We don't need to check if the scorer is null
+                // since we are sure that there are documents to replay (docIdSetIterator it not empty).
+                docIt = scorer.iterator();
+                context.subCollector.setScorer(scorer);
+            }
             while ((docID = docIdSetIterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                if (needsScores) {
+                    if (docIt.docID() < docID) {
```

Comment:
this could be an assertion I think: `assert docIt.docID() < docID`? the scorer can't match fewer documents than docIdSetIterator?